contracts/PeeranhaCommunityToken.sol

```diff
@@ -0,0 +1,126 @@
+    uint256 maxRewardPerUser;
+    uint256 sumAccruedTokens;
+    uint256 sumSpentTokens;
+    uint256 createTime;
+    mapping(uint16 => uint256) periodPool;
+    address peeranhaCommunityTokenFactoryAddress;     // address or interface?
+  }
+
+  CommunityToken communityToken;
+
+  constructor(string memory name, string memory symbol, address contractAddress, uint256 maxRewardPerPeriod, uint256 activeUsersInPeriod, address peeranhaCommunityTokenFactoryAddress) {
+    communityToken.name = name;
+    communityToken.symbol = symbol;
+    communityToken.contractAddress = contractAddress;
+    communityToken.maxRewardPerPeriod = maxRewardPerPeriod;
+    communityToken.activeUsersInPeriod = activeUsersInPeriod;
+    communityToken.maxRewardPerUser = maxRewardPerPeriod / activeUsersInPeriod;
+    communityToken.createTime = CommonLib.getTimestamp();   // how set?
+    communityToken.peeranhaCommunityTokenFactoryAddress = peeranhaCommunityTokenFactoryAddress;
+  }
+
+  // This is to support Native meta transactions
+  // never use msg.sender directly, use _msgSender() instead
+  function _msgSender()
+      internal
+      override(ChildMintableERC20Upgradeable)
+      view
+      returns (address sender)
+  {
+      return ChildMintableERC20Upgradeable._msgSender();
+  } // deleted override(ContextUpgradeable) !!
+
+  function getBalance() public view override returns(uint256) {   // public?
+    uint256 balance;
+    if (communityToken.contractAddress == address(0)) {   // native token
+      balance = address(this).balance;
+    } else {
+      balance = IERC20Upgradeable(communityToken.contractAddress).balanceOf(address(this));
+    }
+
+    require(balance >= communityToken.sumSpentTokens, "error_balance");
+    return balance - communityToken.sumSpentTokens;
+  }
+
+  // get pool
+  function getTotalPeriodReward(uint16 period) private view returns(uint256) {
+    return communityToken.periodPool[period];
+  }
+
+  // set pool
+  function setTotalPeriodReward(RewardLib.PeriodRewardShares memory periodRewardShares, uint16 period) external override {
+    require(_msgSender() == communityToken.peeranhaCommunityTokenFactoryAddress, "only_community_token_factory_can_call_this_action"); // -> _msgSender(); ?
+
+    uint256 totalPeriodReward = reduceRewards(communityToken.maxRewardPerPeriod * FRACTION, period);
+    if (periodRewardShares.activeUsersInPeriod.length <= communityToken.activeUsersInPeriod) {
+      uint256 maxPeriodRewardPerUser = periodRewardShares.activeUsersInPeriod.length * communityToken.maxRewardPerUser * FRACTION;   // min?
+      totalPeriodReward = CommonLib.minUint256(totalPeriodReward, maxPeriodRewardPerUser);
+    }
+    totalPeriodReward = CommonLib.minUint256(totalPeriodReward, getBalance());
+    
+    communityToken.sumSpentTokens += totalPeriodReward;
+    communityToken.periodPool[period] = totalPeriodReward;
+  }
+
+  function reduceRewards(uint256 rewardPeriod, uint16 period) private pure returns(uint256) {
```

Comment:
We are not reducting rewards for communities but we need to provide an action to update reward settings.